### ISSUE TITLE
Quadro support

### ISSUE THREAD

**Issue by cosandr** [shred/pyquaero#3]
Hey,

First off, thanks for making this library, works great with my 6 LT. I was interested in getting it to work with a Quadro as well, but I don't know how any of this USB stuff works. I was trying to get a dump [(see my fork](https://github.com/cosandr/pyquaero/tree/quadro), but it fails in usb.py at send_report with `[Errno 2] Entity not found`. The Quadro has one less interface (?), this is why I changed the range in `__init__`.

If this is something you might be willing to look into, I'd gladly help you however I can. I can probably figure out how to read the firmware (I'm only interested in reading the fan RPM/temperature sensors) but this USB stuff is way over my head for now. 

<details>
    <summary>lsusb -v -d 0c70:f00d</summary>

```
Bus 001 Device 002: ID 0c70:f00d MCT Elektronikladen QUADRO
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.10
  bDeviceClass          239 Miscellaneous Device
  bDeviceSubClass         2
  bDeviceProtocol         1 Interface Association
  bMaxPacketSize0        64
  idVendor           0x0c70 MCT Elektronikladen
  idProduct          0xf00d
  bcdDevice            1.00
  iManufacturer           1 Aqua Computer GmbH & Co. KG
  iProduct                2 QUADRO
  iSerial                 3 19349-40074
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0041
    bNumInterfaces          2
    bConfigurationValue     1
    iConfiguration          4 Default Configuration
    bmAttributes         0x01
      (Missing must-be-set bit!)
      (Bus Powered)
    MaxPower                2mA
    Interface Association:
      bLength                 8
      bDescriptorType        11
      bFirstInterface         0
      bInterfaceCount         1
      bFunctionClass        255 Vendor Specific Class
      bFunctionSubClass       0
      bFunctionProtocol       0
      iFunction               2 QUADRO
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass      0
      bInterfaceProtocol      0
      iInterface              2 QUADRO
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      0
      iInterface              5 Device
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      60
          Report Descriptor: (length is 60)
            Item(Global): Usage Page, data= [ 0x00 0xff ] 65280
                            (null)
            Item(Local ): Usage, data= [ 0x01 ] 1
                            (null)
            Item(Main  ): Collection, data= [ 0x01 ] 1
                            Application
            Item(Global): Report ID, data= [ 0x01 ] 1
            Item(Global): Logical Minimum, data= [ 0x00 ] 0
            Item(Global): Logical Maximum, data= [ 0xff 0x00 ] 255
            Item(Global): Report Size, data= [ 0x08 ] 8
            Item(Global): Report Count, data= [ 0xbb 0x00 ] 187
            Item(Local ): Usage, data= [ 0x01 ] 1
                            (null)
            Item(Main  ): Input, data= [ 0x02 ] 2
                            Data Variable Absolute No_Wrap Linear
                            Preferred_State No_Null_Position Non_Volatile Bitfield
            Item(Global): Report ID, data= [ 0x02 ] 2
            Item(Global): Logical Minimum, data= [ 0x00 ] 0
            Item(Global): Logical Maximum, data= [ 0xff 0x00 ] 255
            Item(Global): Report Size, data= [ 0x08 ] 8
            Item(Global): Report Count, data= [ 0x0a 0x00 ] 10
            Item(Local ): Usage, data= [ 0x01 ] 1
                            (null)
            Item(Main  ): Output, data= [ 0x02 ] 2
                            Data Variable Absolute No_Wrap Linear
                            Preferred_State No_Null_Position Non_Volatile Bitfield
            Item(Global): Report ID, data= [ 0x03 ] 3
            Item(Global): Report Count, data= [ 0xc0 0x03 ] 960
            Item(Local ): Usage, data= [ 0x00 ] 0
                            (null)
            Item(Main  ): Feature, data= [ 0x02 0x01 ] 258
                            Data Variable Absolute No_Wrap Linear
                            Preferred_State No_Null_Position Non_Volatile Buffered Bytes
            Item(Global): Report ID, data= [ 0x08 ] 8
            Item(Global): Report Count, data= [ 0xf4 0x03 ] 1012
            Item(Local ): Usage, data= [ 0x00 ] 0
                            (null)
            Item(Main  ): Feature, data= [ 0x02 0x01 ] 258
                            Data Variable Absolute No_Wrap Linear
                            Preferred_State No_Null_Position Non_Volatile Buffered Bytes
            Item(Main  ): End Collection, data=none
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               5
Binary Object Store Descriptor:
  bLength                 5
  bDescriptorType        15
  wTotalLength       0x0021
  bNumDeviceCaps          1
  Platform Device Capability:
    bLength                28
    bDescriptorType        16
    bDevCapabilityType      5
    bReserved               0
    PlatformCapabilityUUID    {d8dd60df-4589-4cc7-9cd2-659d9e648a9f}
    CapabilityData[0]    0x00
    CapabilityData[1]    0x00
    CapabilityData[2]    0x03
    CapabilityData[3]    0x06
    CapabilityData[4]    0xb2
    CapabilityData[5]    0x00
    CapabilityData[6]    0x07
    CapabilityData[7]    0x00
can't get debug descriptor: Resource temporarily unavailable
Device Status:     0x0000
  (Bus Powered)
```

  </details>

**Comment by shred** [comment]
Actually, the firmware dump is not really helpful for reverse engineering. For the Aquaero board, I installed the aquasuite software. I also installed [Wireshark](https://www.wireshark.org/) to eavesdrop the USB communication between aquasuite and the Aquaero. After that, I changed all the settings in aquasuite individually, and analyzed the USB dumps for the bytes that have been changed. This way, I could figure out how the USB communication is working, and the meaning of the configuration bytes.

Yes, it's actually a lot of work! :smile:

For this reason, it is not really feasible to reverse engineer the communication withouth having the board at hand. And after that, a board is still needed for keeping the library up to date, and for debugging purposes.

So, to be honest, I'm not really interested in adding support for the Quadro board. I'm sorry... But it would consume a lot of my spare time, without reasonable personal benefit.

I'm closing this issue. However, I would gladly merge a pull request if you're successful with your fork, and if you'd also agree to keep the Quadro part up to date.

**Comment by cosandr** [comment]
Hey, I've managed to read most stuff from a Quadro (thanks to Martin from HWiNFO for sharing his structure findings). I'm only 100% sure about the temperature sensors as that's the only thing I have plugged in though.

It seems to behave quite weirdly as well, if I try to read too quick (about 1/second seems to be the limit) it'll split the data in two responses, the length isn't even consistent (sometimes sends no data, sometimes 64 bytes, 128..). It caused me a bit of a headache trying to figure out what was going on (reading the firmware and structure at init was consistently leading to this behavior).

In the end I've decided to just use read_status and thus cache the result between get_firmware and get_structure. read_status will also run another request if the data isn't 188 bytes (I don't know what the length is for FW < 1018) and that works whenever the response is incomplete.

As for writing, I have no plans to work on that, I just want to read data from it for a Prometheus exporter.
